components/fxa-client/android/src/main/java/mozilla/appservices/fxa/State.kt

```diff
@@ -0,0 +1,126 @@
+     *
+     * If there is already an auth flow in progress, this will cancel it and start a new one.
+     * Attempts to pass the state/code from the previous flow to `completeOAuthFlow()` will cause
+     * `FxaException.WrongAuthFlow` to be thrown.
+     *
+     * @param scopes List of OAuth scopes for which the client wants access
+     * @param entryPoint The UI entryPoint used to start this flow. An arbitrary
+     * string which is recorded in telemetry by the server to help analyze the
+     * most effective touchpoints
+     * @return url to navigate the user to
+     * @throws FxaException.InvalidState: the account was not in the `Disconnected` state
+     */
+    abstract suspend fun beginOAuthFlow(
+        entryPoint: String,
+    ): String
+
+    /**
+     * Initiate an OAuth pairing flow and get the URL send the user to
+     *
+     * If there is already an auth flow in progress, this will cancel it and start a new one.
+     * Attempts to pass the state/code from the previous flow to `completeOAuthFlow()` will cause
+     * `FxaException.WrongAuthFlow` to be thrown.
+     *
+     * @param pairingUrl URL string for pairing
+     * @param scopes List of OAuth scopes for which the client wants access
+     * @param entryPoint The UI entryPoint used to start this flow. An arbitrary
+     * string which is recorded in telemetry by the server to help analyze the
+     * most effective touchpoints
+     * @return url to navigate the user to
+     * @throws FxaException.InvalidState: the account was not in the `Disconnected` state
+     */
+    abstract suspend fun beginPairingFlow(
+        pairingUrl: String,
+        entryPoint: String,
+    ): String
+
+    /**
+     * Authenticates the current account using the [code] and [state] parameters obtained via the
+     * OAuth flow initiated by [beginOAuthFlow].
+     *
+     * @param code OAuth code string
+     * @param state state token string
+     * @throws FxaException.InvalidState: the account was not in the `Disconnected` state
+     * @throws FxaException.WrongAuthFlow: the code/state is not valid for the current auth flow
+     */
+    abstract suspend fun completeOAuthFlow(code: String, state: String)
+
+    /**
+     * Cancels any current auth flow
+     *
+     * @throws FxaException.InvalidState: the account was not in the `Disconnected` state
+     */
+    abstract suspend fun cancelOauthFlow()
+
+    /**
+     * Disconnect the account.
+     *
+     * Use this when device record is no longer relevant, e.g. while logging out. On success, other
+     * devices will no longer see the current device in their device lists.
+     *
+     * @throws FxaException.InvalidState: the account was not in the `Connected` state
+     */
+    abstract suspend fun disconnect()
+}
+
+sealed class AccountState  {
+    /** The account is waiting for the `StateManager.initialize()` call */
+    object Uninitialized : AccountState()
```

Comment:
I'm not sure if this is needed, but it seems like a lot of consumers are creating a nullable firefox account instance to defer the disk IO until after startup.  It might be better to formally encode that as a state of the account.